### PR TITLE
Import spec actor types

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Use spec actor types, unifying various go actor types

# Implementation

- Replace `tokenamount.Tokenamount` with TokenAmount from `specs-actors` `abi` package. Use `specs-actors` `big` type for operations
- Remove `types.Signature`, replace with Signature from `specs-actors` `crypto` package. Since this signature intentionally lacks a Verify function, add it as a function provided by the node implementation interface, which makes sense anyway since the node should have its own implementation of signature as it's a security critical feature
- Remove `types.SignedVoucher`, replace with SignedVoucher from `specs-actors` `paych` package
- Move `types.SignedStorageAsk` to `storagemarket` package, since this is the only place it's used and moreover, it's something that's unique to the shared markets as asks are not on chain
- Remove shared folder entirely
- Replace `storagemarket.StorageDealProposal` with `spec-actors` ClientDealProposal + Deal Proposal. This also necessitates switching to real PieceCIDs, which we handle with the `go-fil-commcid` package

### CODE REVIEW

File: storagemarket/impl/client_storagemarket.go [hunk 1]
```diff
@@ -87,21 +86,21 @@ func (c *Client) ProposeStorageDeal(
...
-		PricePerEpoch:      price,
-		ProposalExpiration: uint64(proposalExpiration),
-		Duration:           uint64(duration),
-		Client:             addr,
-		ProviderAddress:    info.Address,
-		MinerWorker:        info.Worker,
-		MinerID:            info.PeerID,
+		Data:            data,
+		PricePerEpoch:   price,
+		StartEpoch:      startEpoch,
```
Comment (shannonwells) [inline storagemarket/impl/client_storagemarket.go]: We'll need to make sure we guard against negative epochs now too.

File: storagemarket/types.go [hunk 2]
```diff
@@ -217,11 +166,14 @@ type StorageProvider interface {
 type StorageProviderNode interface {
 	MostRecentStateId(ctx context.Context) (StateKey, error)
 
+	// Verify a signature against an address + data
+	VerifySignature(signature crypto.Signature, signer address.Address, plaintext []byte) bool
```
Comment (ingar) [inline storagemarket/types.go]: 💯

File: storagemarket/integration_test.go [hunk 3]
```diff
@@ -128,6 +132,121 @@ func TestMakeDeal(t *testing.T) {
 	assert.Equal(t, pd.State, storagemarket.StorageDealActive)
 }
 
+func TestMakeDealOffline(t *testing.T) {
```
Comment (ingar) [inline storagemarket/integration_test.go]: Nice!

File: storagemarket/impl/provider.go [hunk 4]
```diff
@@ -313,18 +312,22 @@ func (p *Provider) ImportDataForDeal(ctx context.Context, propCid cid.Cid, data
...
 	_ = n // TODO: verify n?
 
+	pieceSize := uint64(tempfi.Size())
+
 	_, err = tempfi.Seek(0, io.SeekStart)
 	if err != nil {
 		return xerrors.Errorf("failed to seek through temp imported file: %w", err)
 	}
 
-	commP, err := p.pio.ReadPiece(tempfi)
```
Comment (shannonwells) [inline storagemarket/impl/provider.go]: is p.pio used anymore?

File: storagemarket/impl/client.go [hunk 5]
```diff
@@ -200,22 +202,23 @@ func (c *Client) Start(ctx context.Context, p ClientDealProposal) (cid.Cid, erro
 		return cid.Undef, xerrors.Errorf("computing commP failed: %w", err)
 	}
 
-	dealProposal := &storagemarket.StorageDealProposal{
-		PieceRef:             commP,
-		PieceSize:            uint64(pieceSize),
+	dealProposal := market.DealProposal{
+		PieceCID:             commcid.PieceCommitmentV1ToCID(commP),
```
Comment (ingar) [inline storagemarket/impl/client.go]: Does it make sense to have `PieceIO.GeneratePieceCommitment` return a `cid.CID`?  Would remove a little boilerplate from client and provider.

File: storagemarket/impl/provider_states.go [hunk 6]
```diff
@@ -45,13 +46,13 @@ func (p *Provider) validating(ctx context.Context, deal MinerDeal) (func(*MinerD
 	if err != nil {
 		return nil, err
 	}
-	if head.Height() >= deal.Proposal.ProposalExpiration {
+	if head.Height() >= deal.Proposal.StartEpoch {
```
Comment (ingar) [inline storagemarket/impl/provider_states.go]: Not really related to this PR, but this check doesn't give us any slack time to seal/commit a sector before the `StartEpoch`.

File: storagemarket/impl/client.go (hunk 5 above)
Comment (hannahhoward) [reply storagemarket/impl/client.go]: hmmm... good question. Let's leave it for a future discussion

File: storagemarket/impl/provider_states.go (hunk 6 above)
Comment (hannahhoward) [reply storagemarket/impl/provider_states.go]: yea we should make decision about that in the future

File: retrievalmarket/impl/clientstates/client_states_test.go [hunk 7]
```diff
@@ -474,22 +475,22 @@ func TestProcessNextResponse(t *testing.T) {
...
+var defaultPaymentRequested = abi.NewTokenAmount(500000)
 
 func makeDealState(status retrievalmarket.DealStatus) *retrievalmarket.ClientDealState {
 	return &retrievalmarket.ClientDealState{
 		TotalFunds:       defaultTotalFunds,
 		MinerWallet:      address.TestAddress,
 		ClientWallet:     address.TestAddress2,
 		PayCh:            address.TestAddress2,
-		Lane:             uint64(10),
+		Lane:             int64(10),
```
[the remark this answers is not in the document]
Comment (magik6k) [reply retrievalmarket/impl/clientstates/client_states_test.go]: Wait, that's likely a spec bug

File: retrievalmarket/types.go [hunk 8]
```diff
@@ -187,14 +187,14 @@ type RetrievalProvider interface {
...
-	SetPricePerUnseal(price tokenamount.TokenAmount)
+	SetPricePerUnseal(price abi.TokenAmount)
 	ListDeals() map[ProviderDealID]ProviderDealState
 }
 
 // RetrievalProviderNode are the node depedencies for a RetrevalProvider
 type RetrievalProviderNode interface {
-	UnsealSector(ctx context.Context, sectorId uint64, offset uint64, length uint64) (io.ReadCloser, error)
-	SavePaymentVoucher(ctx context.Context, paymentChannel address.Address, voucher *types.SignedVoucher, proof []byte, expectedAmount tokenamount.TokenAmount) (tokenamount.TokenAmount, error)
+	UnsealSector(ctx context.Context, sectorID uint64, offset uint64, length uint64) (io.ReadCloser, error)
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply retrievalmarket/types.go]: seriously.

File: storagemarket/impl/provider.go (hunk 4 above)
Comment (hannahhoward) [reply storagemarket/impl/provider.go]: it is. elsewhere

File: retrievalmarket/impl/clientstates/client_states_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (magik6k) [reply retrievalmarket/impl/clientstates/client_states_test.go]: I'd expect this to be changed back to uint64 in the spec

[the remark this answers is not in the document]
Comment (magik6k) [reply retrievalmarket/impl/clientstates/client_states_test.go]: This being int64 is a spec bug

[the remark this answers is not in the document]
Comment (hannahhoward) [reply retrievalmarket/impl/clientstates/client_states_test.go]: yea I dunno what to do about this. I wish we weren't using signed ints in spec in so many places.

File: storagemarket/impl/client_storagemarket.go (hunk 1 above)
Comment (hannahhoward) [reply storagemarket/impl/client_storagemarket.go]: I feel like the check to see if StartEpoch > current chain height protects against that, but I wish we were using unsigned ints in spec

File: retrievalmarket/impl/clientstates/client_states_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (magik6k) [reply retrievalmarket/impl/clientstates/client_states_test.go]: https://github.com/filecoin-project/specs-actors/commit/e2abd13ec4b1b1ce977c71a6b0389c7a39ee487e

**Pull request merged**